docs/spec/lightclient/detection.md

```diff
@@ -0,0 +1,491 @@
+header that was not generated by Tendermint consensus.
+
+This specification, the fork detector, is a "second line of defense",
+in case the 1/3 assumption is violated. Its goal is to detect fork (conflicting headers) and collect
+evidence. However, it is impractical to probe all full nodes. At this
+time we consider a simple scheme of maintaining an address book of
+known full nodes from which a small subset (e.g., 4) are chosen
+initially to communicate with. More involved book keeping with
+probabilistic guarantees can be considered at later stages of the
+project.
+
+The light client maintains a simple address book containing addresses
+of full nodes that it can pick as primary and secondaries.  To obtain
+a new header, the light client first does [verification](verification)
+with the primary, and then cross-checks the header with the
+secondaries using this specification.
+
+
+### Tendermint Consensus and Forks
+
+#### **[TMBC-GENESIS]**
+Let *Genesis* be the agreed-upon initial block (file).
+
+#### **[TMBC-FUNC]**
+> **TODO** be more precise. +2/3 of b.NextV = c.Val signed c. For now
+> the following will do:
+
+Let b and c be two light blocks
+with *b.Header.Height + 1 = c.Header.Height*. We define **signs(b,c)**
+    iff `ValidAndVerified(b, c)` 
+
+> **TODO** be more precise. +1/3 of b.NextV signed c. For now
+> the following will do:
+
+Let *b* and *c* be two light blocks. We define **supports(b,c,t)** 
+    iff `ValidAndVerified(b, c)` at time *t*
+
+
+> The following formalizes that *b* was properly generated by
+> Tendermint; *b* can be traced back to genesis
+
+#### **[TMBC-SEQ-ROOTED]**
+Let *b* be a light block. 
+We define *sequ-rooted(b)* iff for all i, 1 <= i < h = b.Header.Height,
+there exist light blocks a(i) s.t.
+   - *a(1) = Genesis* and
+   - *a(h) = b* and
+   - *signs( a(i) , a(i+1) )*.
+
+> The following formalizes that *c* is trusted based on *b* in
+> skipping verification. Observe that we do not require here (yet)
+> that *b* was properly generated.
+
+#### **[TMBC-SKIP-ROOT]**
+Let *b* and *c* be light blocks. We define *skip-root(b,c,t)* if at
+time t there exists an *h* and a sequence *a(1)*, ... *a(h)* s.t.
+   - *a(1) = b* and
+   - *a(h) = c* and
+   - *supports( a(i), a(i+1), t)*, for all i, *1 <= i < h*.
+
+> **TODO** In the above we might use a sequence of times t(i). Not sure.
+
+> **TODO:** I believe the following definition
+> corresponds to **Slashable fork** in
+> [forks][tendermintfork]. Please confirm!  
+> Observe that sign-skip-match  is even defined if there is a fork on
+> the chain.
+
+#### **[TMBC-SIGN-SKIP-MATCH]**
+Let *a*, *b*, *c*, be light blocks and *t* a time, we define 
+*sign-skip-match(a,b,c,t) = true* iff
+   - *sequ-rooted(a)* and
+   - *sequ-rooted(b)* and
+   - *b.Header.Height = c.Header.Height* and
+   - *skip-root(a,c,t)*
+
+implies *b = c*.
+
+If there exists three light blocks a, b, and c, with 
+*sign-skip-match(a,b,c,t) =
+false* then we have a *slashable fork*.
```

Comment:
I am wondering if this is too strong. Isn't it sufficient to have skip-root(a,b,t) and skip-root(a,c,t) and b.height = c.height to have a slashable fork? Or we need to have also a block d such that d.height = c.height and sequ-rooted(d)?

Reply:
Yes! Good catch!!